rfcs/20190225-tf-on-demand.md

```diff
@@ -0,0 +1,191 @@
+Downloader will now detect your system:
+	- Detecting CPU…….  Found Intel Core i7-8700K
+	- Detecting GPU……. Found NVIDIA GTX 980
+	- Detecting CUDA…… Found 9.2
+	- Detecting cuDNN….. Found 7.4
+	- Detecting Distribution……. Found Ubuntu 18.04
+
+Requesting TF build with options:
+    CPU options: -mavx2
+    GPU enabled: yes
+    CUDA version: 9.2
+    cuDNN version: 7.4
+    CUDA compute capability: 5.2
+    GCC version: 7.3
+```
+
+
+It is possible that the system this script is run on may be incompatible for running TF. Such incapabilities and what the script is able to do can be categorized into two failure modes:
+
+
+
+*   *Hard Constraint Failures*
+    *   32-bit OS
+    *   GPU too old (Cuda Compute Capability older than 3.0)
+*   *Soft Constraint Failures*
+    *   CUDA version too old
+    *   cuDNN missing
+    *   Other runtime libraries missing or too old (libstdc++)
+    *   Unsupported python version found
+
+In hard constraint failure cases, the tool will point users to cloud options. In soft constraint failure cases, the tool will offer alternative build configurations (CPU only) or request users to install the missing software.
+
+We propose distributing this binary through PyPI. However, please note that binary distribution method is still under discussion.
+
+### Back-end
+
+**TF Builder Service**
+
+TF builder service will require a proto with build options. This proto will define all options the system recognizes to build TF. Once the request is received, the system will check the cache to see if we already have such a package built. If yes, then it will send back the URL for downloading this package. A sample proto (not final) is shown below:
+
+```
+proto build_options {
+   string version = 0;
+   enum CpuOptions {
+       SSE3 = 0;
+       SSE4 = 1;
+       SSE4_1 = 2;
+       SSE4_A = 3;
+       SSE4_2 = 4;
+       AVX = 5;
+       AVX2 = 6;
+       AVX512F = 7;
+   }
+   repeated CpuOptions cpu_options = 1;
+
+   // CUDA options
+   bool cuda_enabled = 2;
+   enum NvidiaGPUGeneration {
+       KEPLER = 0;
+       MAXWELL = 1;
+       PASCAL = 2;
+       VOLTA = 3;
+       TURING = 4;
+   }
+   NvidiaGPUGeneration gpu_generation = 3;
+   string CUDA_version = 4;
+   string cuDNN version = 5;
+……
+   // Free formed string of options to append to bazel.
+   string extra_options = 100;
+}
+```
+
+If no such package exists in cache, then the system will execute the following commands with the appropriate flags and environment variables to newly build it.
+
+```
+git checkout <tag>
+yes "" | ./configure
+bazel build tensorflow/tools/pip_package:build_pip_package
+./bazel-bin/tensorflow/tools/pip_package/build_pip_package
+```
+
+Once the build is complete, the package will be stored in the cache before sending out a link to the user for download.
+
+**Package Cache**
+
+We will design the package cache as a simple GCS bucket supported by a simple database. The database can have the following schema: 
```

Comment:
How long will the built artifacts be available for download? Will it be possible to list available builds (e.g. have listing of `gs://tensorflow-artifacts/on-demand/`?

IMHO its worth to consider keeping build options somewhere handy (e.g. publish well defined proto or JSON/YAML alongside with the resulting wheel files) so that a user can see build configuration of a wheel file before downloading it or cross-checking it. This is also handy for automated systems that would like to consume such metadata when using built wheel files.

Another option (instead of creating a metadata files) is to keep prefix consisting of build configuration options  as we did originally [here](http://tensorflow.pypi.thoth-station.ninja/index/) the issue is in maintaining such prefix and keeping it parse-able.

Reply:
@hyeygit ?

Reply:
For now, the built artifacts will be available indefinitely. In the future, however, we may think about cleaning them up. 

Also, thank you for the suggestions. We will definitely consider the options.

Reply:
For now, we plan to have the built artifacts be available indefinitely. In the future, however, we may think about cleaning them up.

Reply:
Regarding storing the build options, we are still actively discussing, trying to come up with a solution that would be manageable and scalable. The second option you provided has actually been brought up. I do agree that giving the users the option to retrieve the build configuration of a wheel file would be handy. Will make sure to consider the first option.